### ISSUE TITLE
Panic due to NPE when processing WiredList

### ISSUE THREAD

**Issue by justinjc** [m3db/m3#2573]
Stack trace:
```
panic: runtime error: invalid memory address or nil pointer dereference

github.com/m3db/m3/src/x/ident.BytesID.Equal(...)
    github.com/m3db/m3/src/x/ident/bytes_id.go:46
github.com/m3db/m3/src/dbnode/storage/series.(*dbSeries).OnEvictedFromWiredList(...)
    github.com/m3db/m3/src/dbnode/storage/series/series.go:580
github.com/m3db/m3/src/dbnode/storage.(*dbShard).OnEvictedFromWiredList(...)
    github.com/m3db/m3/src/dbnode/storage/shard.go:532
github.com/m3db/m3/src/dbnode/storage/block.(*WiredList).insertAfter(...)
    github.com/m3db/m3/src/dbnode/storage/block/wired_list.go:299
github.com/m3db/m3/src/dbnode/storage/block.(*WiredList).pushBack(...)
    github.com/m3db/m3/src/dbnode/storage/block/wired_list.go:348
github.com/m3db/m3/src/dbnode/storage/block.(*WiredList).processUpdateBlock(...)
    github.com/m3db/m3/src/dbnode/storage/block/wired_list.go:239
github.com/m3db/m3/src/dbnode/storage/block.(*WiredList).Start.func1(...)
    github.com/m3db/m3/src/dbnode/storage/block/wired_list.go:169
```

~~This is a regression after an upgrade from:~~
~~b429452f1fad0533c5b3a0b20dd23d1963708203 (from v0.14.2)~~
~~to:~~
~~99477fc131a373b8d226dfd4bb42795f13c9ba0c (from v0.15.12)~~
**Update**: This bug appears to have been present for a very long time.

The stack trace indicates that the `id` in the series is nil. Judging purely from the comment block above the `id` field:
https://github.com/m3db/m3/blob/99477fc131a373b8d226dfd4bb42795f13c9ba0c/src/dbnode/storage/series/series.go#L64-L68
I'm guessing that this might be related to this commit: https://github.com/m3db/m3/commit/b4be7b1356d72a4052ecf21c6d027cad5a027b07

cc: @robskillington

**Comment by robskillington** [comment]
Hm @justinjc I think this actually might be a race case where a series is evicted from the wired list and it is also evicted from the shard map and therefore `s.id` is nil on the series.

Could you perhaps try adding a `closed bool` to the series and checking that at the beginning of "OnEvictedFromWiredList" before checking against the ID?  

If that works we may want to create this concept of open or closed, and then set it to false on "Reset(...)" and then set it true on "Close()".

This will help us discern whether it's a valid use of the series or if it's simply not in an open state and used outside of the shard lock map (like the case here where the series is being used outside of the RLock/RUnlock being held of the shard map).

**Comment by justinjc** [comment]
Quick update: it appears that this bug is present in older versions too (pre-v0.15.x). We seemed to have encountered this in clusters with old builds, but for some reason it procs significantly more frequently with new builds.

I've attempted to repro this with [this test](https://github.com/m3db/m3/pull/3056/files#diff-b13338c27395c14f1eb7ba8eca68508feb2941e3ad6c00db9c20601c2b7cc2bb), which continuously writes and flushes series so that the series gets evicted (closed), nil-ing the ID. At the same time, I fetch IDs in a loop to continuously evict series from the wired list. However, I have yet to make it repro with this.
